### ISSUE TITLE
Enforce child entity concrete type regarding the parent concrete type in a TPH scenario

### ISSUE THREAD

**Issue by orty** [dotnet/efcore#33332]
Hi,

We currently use EF Core 8 in our project. We have the following entities in our DB :

```c#
public abstract class EntityA
{
    public int Id { get; set; }
    public ICollection<EntityB> EntityBs { get; set; }
    public abstract EntityType Type { get; set; }
    // Common properties
}

public class EntityASubtype1 : EntityA
{
     public override EntityType Type { get; set; } = EntityType.Subtype1;
    // Subtype specific properties
}

public class EntityASubtype2 : EntityA
{
     public override EntityType Type { get; set; } = EntityType.Subtype2;
    // Subtype specific properties
}

public abstract class EntityB
{
    public int Id { get; set; }
    public int EntityAId { get; set; }
    public abstract EntityType Type { get; set; }
    // Common properties
}

public class EntityBSubtype1 : EntityB
{
     public override EntityType Type { get; set; } = EntityType.Subtype1;
    // Subtype specific properties
}

public class EntityBSubtype2 : EntityB
{
     public override EntityType Type { get; set; } = EntityType.Subtype2;
    // Subtype specific properties
}
```

And we implemented a TPH strategy using the enum `EntityType` as a discriminator. Each concrete `EntityA` type is, by design in our business need, bound to one and only one concrete `EntityB` type, and they share the same discriminator value. 

We tried to use the the discriminator from `EntityA` in the `DbContext` declaration for `EntityB` and apparently this is not supported.
We then tried to use the `EntityA.Type` column as a part of the foreign key in the `EntityB` declaration, which leads to a composite foreign key made like so:
```#
.HasPrincipalKey(t => new { entityA.Id, entityA.Type })
.HasForeignKey(ttv => new { entityB.EntityAId, entityB.Type })
```
as an attempt to "force" the Type discriminator value on the child entity to its value on the parent entity. This is closest we got from enforcing the child concrete type after the parent concrete type.

Now, when we query the database for `EntityA`s, including their `EntityB`s (there is also OData and AutoMapper in the mix, but I still hope that there is something to do on the EF side), EF generates several 
```SQL
SELECT CASE WHEN ... THEN
```
statements, one for each value of the `EntityA.Type` discriminator, which is understandable for the later-to-happen materialization of the concrete entities, but also generates the same statements for the child entities, and worse, a `LEFT JOIN` on the `EntityBs` table for each of these values, which leads to a exponentially long SQL Query, which in turn can go from a 30 rows expected result to 100.000+ rows.

Long story short, is there any way to enforce the type of a child entity in relation to its parent when both entities implement a TPH strategy ?

### Include provider and version information

EF Core version: 8
Database provider: (e.g. Microsoft.EntityFrameworkCore.SqlServer)
Target framework: (e.g. .NET 8.0)

**Comment by AndriySvyryd** [comment]
There are some options that would allow to represent this mapping.

In C# you could enforce this by making `EntityA` generic:
```C#
public abstract class EntityA<TB>
    where TB : EntityB
{
    public int Id { get; set; }
    public ICollection<TB> EntityBs { get; set; }
    public abstract EntityType Type { get; set; }
    // Common properties
}
```

If you are unable to do it, it's also possible to configure the relationships between the derived types instead of the base types:
```C#
protected override void OnModelCreating(ModelBuilder modelBuilder)
{
    modelBuilder.Entity<EntityA>().Ignore(e => e.EntityBs);
    modelBuilder.Entity<EntityASubtype1>().HasMany(e => (ICollection<EntityBSubtype1>)e.EntityBs).WithOne();
    modelBuilder.Entity<EntityASubtype2>().HasMany(e => (ICollection<EntityBSubtype2>)e.EntityBs).WithOne();
}
```
This wouldn't enforce the discriminators to be in sync, rather it would let EF assume that they are.

Note that the cast is only used for configuration purposes, it won't be performed at runtime.
